File: androidApp/src/main/java/org/idos/app/ui/screens/credentials/CredentialsViewModel.kt
package org.idos.app.ui.screens.credentials

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import org.idos.app.data.model.Credential
import org.idos.app.data.repository.CredentialsRepository
import org.idos.app.ui.screens.base.BaseViewModel
import timber.log.Timber

// Events that can be triggered from the UI
sealed class CredentialsEvent {
    object LoadCredentials : CredentialsEvent()
    data class CredentialSelected(val credential: Credential) : CredentialsEvent()
    object ClearError : CredentialsEvent()
}

// UI State for the Credentials screen
data class CredentialsUiState(
    val isLoading: Boolean = false,
    val credentials: List<Credential> = emptyList(),
    val error: String? = null,
)

class CredentialsViewModel(
    private val credentialsRepository: CredentialsRepository
) : BaseViewModel<CredentialsUiState, CredentialsEvent>() {

    override fun initialState(): CredentialsUiState = CredentialsUiState()

    override fun onEvent(event: CredentialsEvent) {
        when (event) {
            is CredentialsEvent.LoadCredentials -> loadCredentials()
            is CredentialsEvent.CredentialSelected -> onCredentialSelected(event.credential)
            is CredentialsEvent.ClearError -> clearError()
        }
    }

    private fun loadCredentials() {
        viewModelScope.launch {
            try {
                updateState { copy(isLoading = true, error = null) }
                credentialsRepository.getCredentials().collect { credentials ->
                    Timber.d("got credentials {$credentials}")
                    updateState { copy(credentials = credentials, isLoading = false) }
                }
            } catch (e: Exception) {
                updateState {
                    copy(
                        isLoading = false,
                        error = "Failed to load credentials: ${e.message}"
                    )
                }
            }
        }
    }

    private fun onCredentialSelected(credential: Credential) {
        // Handle credential selection
        // This could navigate to a detail screen or perform an action
    }

    private fun clearError() {
        updateState { copy(error = null) }
    }
}

File: iosApp/iosApp/Features/Credentials/CredentialsViewModel.swift
import Foundation
import Combine
import idos_sdk

/// Credentials state matching Android's CredentialsState
struct CredentialsState {
    var credentials: [Credential] = []
    var isLoading: Bool = false
    var error: String? = nil
}

/// Credentials events matching Android's CredentialsEvent
enum CredentialsEvent {
    case loadCredentials
    case credentialSelected(Credential)
    case clearError
    case refresh
}

/// CredentialsViewModel matching Android's CredentialsViewModel
class CredentialsViewModel: BaseViewModel<CredentialsState, CredentialsEvent> {
    private let credentialsRepository: CredentialsRepositoryProtocol
    private let navigationCoordinator: NavigationCoordinator

    init(
        credentialsRepository: CredentialsRepositoryProtocol,
        navigationCoordinator: NavigationCoordinator
    ) {
        self.credentialsRepository = credentialsRepository
        self.navigationCoordinator = navigationCoordinator
        super.init(initialState: CredentialsState())
        loadCredentials()
    }

    override func onEvent(_ event: CredentialsEvent) {
        switch event {
        case .loadCredentials, .refresh:
            loadCredentials()
        case .credentialSelected(let credential):
            navigateToDetail(credential)
        case .clearError:
            state.error = nil
        }
    }

    private func loadCredentials() {
        Task { [weak self] in
            guard let self = self else { return }

            await MainActor.run {
                self.state.isLoading = true
                self.state.error = nil
            }

            do {
                let credentials = try await credentialsRepository.getCredentials()
                await MainActor.run {
                    self.state.credentials = credentials
                    self.state.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.state.error = "Failed to load credentials: \(error.localizedDescription)"
                    self.state.isLoading = false
                }
            }
        }
    }

    private func navigateToDetail(_ credential: Credential) {
        navigationCoordinator.navigate(to: .credentialDetail(credentialId: credential.id))
    }
}
